### ISSUE TITLE
Test errors on Windows with hdf5-1.10.3

### ISSUE THREAD

**Issue by cgohlke** [h5py/h5py#1088]
Just reporting: I get two test failures with current master, built against hdf5-1.10.3 on Windows:

```
Summary of the h5py configuration
---------------------------------

h5py    2.8.0.post0
HDF5    1.10.3
Python  3.7.0 (v3.7.0:1bf9cc5093, Jun 27 2018, 04:59:51) [MSC v.1914 64 bit (AMD64)]
sys.platform    win32
sys.maxsize     9223372036854775807
numpy   1.14.5
```

```
============================= test session starts =============================
platform win32 -- Python 3.7.0, pytest-3.7.2, py-1.5.4, pluggy-0.7.1
plugins: xdist-1.22.5, pep8-1.0.6, forked-0.2, faulthandler-1.5.0, cov-2.5.1, palladium-1.2.0, hypothesis-3.69.1, celery-4.2.1
collected 492 items

h5py\tests\hl\test_attribute_create.py ..                                [  0%]
h5py\tests\hl\test_dataset_getitem.py .................................. [  7%]
......x........x....x......................x...                          [ 16%]
h5py\tests\hl\test_dataset_swmr.py ssss.F........                        [ 19%]
h5py\tests\hl\test_datatype.py ........                                  [ 21%]
h5py\tests\hl\test_dims_dimensionproxy.py .                              [ 21%]
h5py\tests\hl\test_file.py .......                                       [ 22%]
h5py\tests\hl\test_filters.py ..                                         [ 23%]
h5py\tests\hl\test_threads.py ..                                         [ 23%]
h5py\tests\hl\test_vds\test_highlevel_vds.py .....                       [ 24%]
h5py\tests\hl\test_vds\test_lowlevel_vds.py ...                          [ 25%]
h5py\tests\hl\test_vds\test_virtual_source.py ................           [ 28%]
h5py\tests\old\test_attrs.py ..............                              [ 31%]
h5py\tests\old\test_attrs_data.py ....................                   [ 35%]
h5py\tests\old\test_base.py .....                                        [ 36%]
h5py\tests\old\test_dataset.py .s.......x............................... [ 44%]
.................................x..........................             [ 57%]
h5py\tests\old\test_datatype.py ..                                       [ 57%]
h5py\tests\old\test_dimension_scales.py ....s................            [ 61%]
h5py\tests\old\test_file.py .............ss.ss..................s....... [ 70%]
.                                                                        [ 70%]
h5py\tests\old\test_file_image.py ..                                     [ 71%]
h5py\tests\old\test_group.py ....................F............ssssss.... [ 80%]
...........................ss..................                          [ 89%]
h5py\tests\old\test_h5.py .....                                          [ 90%]
h5py\tests\old\test_h5d_direct_chunk_write.py .                          [ 90%]
h5py\tests\old\test_h5f.py .....                                         [ 91%]
h5py\tests\old\test_h5p.py ......                                        [ 93%]
h5py\tests\old\test_h5t.py ....                                          [ 93%]
h5py\tests\old\test_objects.py ...                                       [ 94%]
h5py\tests\old\test_selections.py ....                                   [ 95%]
h5py\tests\old\test_slicing.py .......................                   [100%]

================================== FAILURES ===================================
_____________ TestDatasetSwmrRead.test_force_swmr_mode_on_raises ______________

self = <h5py.tests.hl.test_dataset_swmr.TestDatasetSwmrRead testMethod=test_force_swmr_mode_on_raises>

    def test_force_swmr_mode_on_raises(self):
        """ Verify when reading a file cannot be forcibly switched to swmr mode.
            When reading with SWMR the file must be opened with swmr=True."""
        with self.assertRaises(ValueError):
>           self.f.swmr_mode = True

X:\Python37\lib\site-packages\h5py\tests\hl\test_dataset_swmr.py:75:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
h5py\_objects.pyx:54: in h5py._objects.with_phil.wrapper
    ???
h5py\_objects.pyx:55: in h5py._objects.with_phil.wrapper
    ???
X:\Python37\lib\site-packages\h5py\_hl\files.py:277: in swmr_mode
    self.id.start_swmr_write()
h5py\_objects.pyx:54: in h5py._objects.with_phil.wrapper
    ???
h5py\_objects.pyx:55: in h5py._objects.with_phil.wrapper
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   TypeError: Unable to convert file format (no write intent on file)

h5py\h5f.pyx:538: TypeError
______________________________ TestLen.test_exc _______________________________

self = <h5py.tests.old.test_group.TestLen testMethod=test_exc>

    def test_exc(self):
        """ len() on closed group gives ValueError """
        self.f.close()
        with self.assertRaises(ValueError):
>           len(self.f)

X:\Python37\lib\site-packages\h5py\tests\old\test_group.py:302:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
h5py\_objects.pyx:54: in h5py._objects.with_phil.wrapper
    ???
h5py\_objects.pyx:55: in h5py._objects.with_phil.wrapper
    ???
X:\Python37\lib\site-packages\h5py\_hl\group.py:385: in __len__
    return self.id.get_num_objs()
h5py\_objects.pyx:54: in h5py._objects.with_phil.wrapper
    ???
h5py\_objects.pyx:55: in h5py._objects.with_phil.wrapper
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   TypeError: Not a location id (invalid object ID)

h5py\h5g.pyx:331: TypeError
========= 2 failed, 465 passed, 19 skipped, 6 xfailed in 4.27 seconds =========
```

**Comment by dotlambda** [comment]
I'm also running into the same 2 failures on x86_64 Linux (NixOS).

**Comment by tacaswell** [comment]
Interesting, it looks like both have changed from `ValueError` to `TypeError`...  At least they are still failing?

Need to dig into this to sort out if we should catch and convert them back to ValueError or let the TypeErrors propgate out...

**Comment by tacaswell** [comment]
I can also reproduce this in linux.

**Comment by tacaswell** [comment]
and on CI on linux

**Comment by mcepl** [comment]
And openSUSE, https://build.opensuse.org/package/live_build_log/openSUSE:Factory/python-h5py/standard/x86_64

**Comment by mcepl** [comment]
Isn't the most brutal [solution](https://github.com/h5py/h5py/files/2464981/valuerror-typerror-failing-tests.patch.txt) the best one?

**Comment by tacaswell** [comment]
@mcepl  I would like to understand _why_ this changed before we just paper over it.

**Comment by tacaswell** [comment]
@epourmal can the hdf5 group comment on these changes and the fix in #1099 ?
